php-frontend/src/main/java/org/sonar/php/metrics/MetricsVisitor.java

```diff
@@ -0,0 +1,118 @@
+  private static final Number[] LIMITS_COMPLEXITY_FUNCTIONS = {1, 2, 4, 6, 8, 10, 12, 20, 30};
+  private static final Number[] FILES_DISTRIBUTION_BOTTOM_LIMITS = {0, 5, 10, 20, 30, 60, 90};
+
+  public static final Kind[] FUNCTION_NODES = {
+    Kind.FUNCTION_DECLARATION,
+    Kind.FUNCTION_EXPRESSION,
+    Kind.METHOD_DECLARATION,
+  };
+
+  public static final Kind[] CLASS_NODES = {
+    Kind.CLASS_DECLARATION,
+    Kind.INTERFACE_DECLARATION,
+    Kind.TRAIT_DECLARATION
+  };
+
+
+  private FileMeasures fileMeasures;
+  private FileLinesContext fileLinesContext;
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    List<Kind> result = new ArrayList<>(Arrays.asList(FUNCTION_NODES));
+    result.addAll(Arrays.asList(CLASS_NODES));
+    result.add(Kind.COMPILATION_UNIT);
+    return result;
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Kind.COMPILATION_UNIT)) {
+      fileMeasures.setFileComplexity(NewComplexityVisitor.complexity(tree));
+
+    } else if (tree.is(Kind.CLASS_DECLARATION)) {
+      fileMeasures.addClassComplexity(NewComplexityVisitor.complexity(tree));
+
+    } else if (tree.is(FUNCTION_NODES)) {
+      fileMeasures.addFunctionComplexity(NewComplexityVisitor.complexityWithoutNestedFunctions(tree));
+    }
+  }
+
+  public FileMeasures getFileMeasures(File file, CompilationUnitTree tree, FileLinesContext fileLinesContext) {
+    this.fileMeasures = new FileMeasures(LIMITS_COMPLEXITY_FUNCTIONS, FILES_DISTRIBUTION_BOTTOM_LIMITS);
+    this.fileLinesContext = fileLinesContext;
+
+    super.analyze(file, tree);
+
+    setCounterMeasures();
+    setLineAndCommentMeasures();
+    return this.fileMeasures;
+  }
+
+  private void setCounterMeasures() {
+    CounterVisitor counter = new CounterVisitor(context().tree());
+    fileMeasures.setClassNumber(counter.getClassNumber());
+    fileMeasures.setFunctionNumber(counter.getFunctionNumber());
+    fileMeasures.setStatementNumber(counter.getStatementNumber());
+  }
+
+  private void setLineAndCommentMeasures() {
+    LineVisitor lineVisitor = new LineVisitor(context().tree());
+    CommentLineVisitor commentVisitor = new CommentLineVisitor(context().tree());
+
+    int linesNumber = lineVisitor.getLinesNumber();
+
+    fileMeasures.setLinesNumber(linesNumber);
+    fileMeasures.setLinesOfCodeNumber(lineVisitor.getLinesOfCodeNumber());
+    fileMeasures.setCommentLinesNumber(commentVisitor.getCommentLineNumber());
+    fileMeasures.setNoSonarLines(commentVisitor.getNoSonarLines());
+
+    Set<Integer> linesOfCode = lineVisitor.getLinesOfCode();
+    Set<Integer> commentLines = commentVisitor.getCommentLines();
+
+    for (int line = 1; line <= linesNumber; line++) {
+      fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, line, linesOfCode.contains(line) ? 1 : 0);
+      fileLinesContext.setIntValue(CoreMetrics.COMMENT_LINES_DATA_KEY, line, commentLines.contains(line) ? 1 : 0);
+    }
+
+    // fixme (is it compliant with our idea of separate logic?)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS1134)